plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RewriteCast.java

```diff
@@ -0,0 +1,112 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.redshift;
+
+import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.plugin.jdbc.expression.AbstractRewriteCast;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.BigintType;
+import io.trino.spi.type.BooleanType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.DoubleType;
+import io.trino.spi.type.IntegerType;
+import io.trino.spi.type.RealType;
+import io.trino.spi.type.SmallintType;
+import io.trino.spi.type.TinyintType;
+import io.trino.spi.type.Type;
+
+import java.sql.Types;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+public class RewriteCast
+        extends AbstractRewriteCast
+{
+    public RewriteCast(BiFunction<ConnectorSession, Type, String> toTargetType)
+    {
+        super(toTargetType);
+    }
+
+    @Override
+    protected boolean eligibleForPushdown(Type sourceType, Type targetType)
+    {
+        if (targetType instanceof TinyintType) {
+            return supportedSourceTypeToCastInIntegralType(sourceType);
+        }
+        if (targetType instanceof SmallintType) {
+            return supportedSourceTypeToCastInIntegralType(sourceType);
+        }
+        if (targetType instanceof IntegerType) {
+            return supportedSourceTypeToCastInIntegralType(sourceType);
+        }
+        if (targetType instanceof BigintType) {
+            return supportedSourceTypeToCastInIntegralType(sourceType);
+        }
+        return false;
+    }
+
+    @Override
+    protected Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type)
+    {
+        if (type instanceof TinyintType) {
+            return Optional.of(new JdbcTypeHandle(Types.TINYINT, Optional.of("tinyint"), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        if (type instanceof SmallintType) {
+            return Optional.of(new JdbcTypeHandle(Types.SMALLINT, Optional.of("smallint"), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        if (type instanceof IntegerType integerType) {
+            return Optional.of(new JdbcTypeHandle(Types.INTEGER, Optional.of(integerType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        if (type instanceof BigintType bigintType) {
+            return Optional.of(new JdbcTypeHandle(Types.BIGINT, Optional.of(bigintType.getBaseName()), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    protected String buildCast(Type sourceType, Type targetType, String expression, String castType)
+    {
+        if (isDecimalType(sourceType) && isIntegralType(targetType)) {
+            return "CAST(ROUND(%s) AS %s)".formatted(expression, castType);
+        }
```

Comment:
I have some doubts if the trino behaviour is a correct one. 
Do you know the internals of the trino behaviour?
Maybe at least add a comment here `to match trino behaviour`? Not sure if it's helpful (kind of obvious if comment exists, and not that obvious if not)

Reply:
In Trino cast from `double` to any other type we use `DoubleOperators` and here is one example where casting from double to long roundup the value.

https://github.com/trinodb/trino/blob/5f191afb5d26fd374f08b61dd58f3e64d70bca00/core/trino-main/src/main/java/io/trino/type/DoubleOperators.java#L165

Anyway, we will not support Float/Double value cast to other types, since for `NaN` and `Infinity` values we are getting behavior deference between with and without pushdown.